### PR TITLE
Apply and Filter Refactor

### PR DESCRIPTION
**Pull request by akleeman**
This moves the `apply` and `filter` methods to be free functions which work with any map like objects.

Note 99% of this is just moved code, the largest change is the need to add a few more template parameters to the free function versions of the formerly class level `apply` and `filter` functions.

### CODE REVIEW

File: include/albatross/src/indexing/apply.hpp [hunk 1]
```diff
@@ -54,6 +40,70 @@ auto apply(const std::vector<ValueType> &xs, const ApplyFunction &f) {
...
+
+template <template <typename...> class Map, typename KeyType,
+          typename ValueType, typename ApplyFunction,
+          typename ApplyType = typename details::value_only_apply_result<
+              ApplyFunction, ValueType>::type,
+          typename std::enable_if<details::is_valid_value_only_apply_function<
+                                      ApplyFunction, ValueType>::value &&
+                                      std::is_same<void, ApplyType>::value,
+                                  int>::type = 0>
+auto apply(const Map<KeyType, ValueType> &map, const ApplyFunction &f) {
```
Comment (jbangelo) [inline include/albatross/src/indexing/apply.hpp]: Return type should probably be explicitly `void`

File: include/albatross/src/indexing/apply.hpp [hunk 2]
```diff
@@ -54,6 +40,70 @@ auto apply(const std::vector<ValueType> &xs, const ApplyFunction &f) {
...
+template <
+    template <typename...> class Map, typename KeyType, typename ValueType,
+    typename ApplyFunction,
+    typename ApplyType = typename details::key_value_apply_result<
+        ApplyFunction, KeyType, ValueType>::type,
+    typename std::enable_if<details::is_valid_key_value_apply_function<
+                                ApplyFunction, KeyType, ValueType>::value &&
+                                std::is_same<void, ApplyType>::value,
+                            int>::type = 0>
+void apply(const Map<KeyType, ValueType> &map, const ApplyFunction &f) {
```
Comment (jbangelo) [inline include/albatross/src/indexing/apply.hpp]: Shouldn't these functions be declared `inline`? Probably true for all of them in this file.

File: include/albatross/src/indexing/filter.hpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+// Map
+
+template <template <typename...> class Map, typename KeyType,
+          typename ValueType, typename ToKeepFunction,
+          typename std::enable_if<details::is_valid_value_only_filter_function<
+                                      ToKeepFunction, ValueType>::value,
+                                  int>::type = 0>
+auto filter(const Map<KeyType, ValueType> &map, ToKeepFunction to_keep) {
```
Comment (jbangelo) [inline include/albatross/src/indexing/filter.hpp]: `inline`?

File: include/albatross/src/indexing/apply.hpp (hunk 2 above)
Comment (akleeman) [reply include/albatross/src/indexing/apply.hpp]: Yes!  Nice catch.

File: include/albatross/src/indexing/apply.hpp (hunk 1 above)
Comment (akleeman) [reply include/albatross/src/indexing/apply.hpp]: good call

**Pull request merged**